docs/graph/dominator-tree.md

```diff
@@ -0,0 +1,718 @@
+
+根据定理 1 我们便可以求出每个点的半支配点了。不难发现计算半支配点的复杂度瓶颈在第二种情况上，我们考虑利用带权并查集优化，每次路径压缩时更新最小值即可。
+
+```c++
+void dfs(int u) {
+  dfn[u] = ++dfc;
+  pos[dfc] = u;
+  for (int i = h[0][u]; i; i = e[i].x) {
+    int v = e[i].v;
+    if (!dfn[v]) {
+      dfs(v);
+      fth[v] = u;
+    }
+  }
+}
+
+int find(int x) {
+  if (fa[x] == x) {
+    return x;
+  }
+  int tmp = fa[x];
+  fa[x] = find(fa[x]);
+  if (dfn[sdm[mn[tmp]]] < dfn[sdm[mn[x]]]) {
+    mn[x] = mn[tmp];
+  }
+  return fa[x];
+}
+
+void getsdom() {
+  dfs(1);
+  for (int i = 1; i <= n; ++i) {
+    mn[i] = fa[i] = sdm[i] = i;
+  }
+  for (int i = dfc; i >= 2; --i) {
+    int u = pos[i], res = INF;
+    for (int j = h[1][u]; j; j = e[j].x) {
+      int v = e[j].v;
+      if (!dfn[v]) {
+        continue;
+      }
+      find(v);
+      if (dfn[v] < dfn[u]) {
+        res = std::min(res, dfn[v]);
+      } else {
+        res = std::min(res, dfn[sdm[mn[v]]]);
+      }
+    }
+    sdm[u] = pos[res];
+    fa[u] = fth[u];
+  }
+}
+
+```
+
+#### 求解直接支配点
+
+##### 转化为 DAG
+
+可是我还是不知道半支配点有什么用！
+
+我们考虑在 $T$ 上对每一个 $u$ 加入 $sdom(u) \rightarrow u$ 的有向边。根据引理 9，新得到的这张图 $G$ 一定是有向无环图；又根据引理 10，我们还发现这样加边不会改变支配关系，因此我们把原图转化为了一张 DAG，利用上文的算法求解即可。
+
+##### 通过半支配点求解
+
+建一堆图也太不优雅了！
+
+**定理 2：** 对于任意节点 $u$，若 $T$ 上从 $sdom(u)$ 到 $w$ 的路径上的任意节点 $v$ 都满足 $sdom(v)\ge sdom(w)$，则 $idom(u) =sdom(u)$。
+
+**证明：** 根据引理 10 我们知道 $idom(u)$ 是 $sdom(u)$ 或其祖先，因此只需证明 $sdom(u) \ dom \ u$。
+
+考虑任意一条 $s$ 到 $u$ 的路径 $P$，我们需要证明 $sdom(u)$ 一定在 $P$ 中。令 $v$ 为 $P$ 中最后一个满足 $x<sdom(u)$ 的节点。如果 $v$ 不存在则必有 $sdom(u)=idom(u) =s$，否则令 $w$ 是 $P$ 中 $v$ 之后在 DFS 树中从 $sdom(u)$ 到 $u$ 的路径上的第一个点。
+
+我们接下来证明 $sdom(w)\le v <sdom(v)$。考虑 $T$ 上 $v$ 到 $w$ 的路径 $v = v_0 \rightarrow \dots v_k = w$，若不成立，则存在 $i\in[1, k- 1], v_i < w$。此时一定存在某个 $j\in [i, k - 1]$ 满足 $v_j$ 是 $w$ 的祖先。由 $v$ 的取值可知 $sdom(u)\le v_j$，于是 $v_j$ 也在 DFS 树中从 $sdom(u)$ 到 $u$ 的路径上，与 $w$ 的定义矛盾，因此 $sdom(w)\le v < sdom(v)$，结合定理的条件有 $y=sdom(u)$，即路径 $P$ 包含 $sdom(u)$。
+
+**定理 3：** 对应任意节点 $u$，$T$ 上从 $sdom(u)$ 到 $u$ 的路径上的所有节点中最小的节点 $v$ 一定满足 $sdom(v)\le sdom(u)$ 和 $idom(v) = idom(u)$。
+
+**证明：** 考虑到 $u$ 本身也满足 $v$ 的条件，因此 $sdom(v)\le sdom(u)$。
+
+由于 $idom(u)$ 是 $v$ 在 $T$ 上的祖先，由引理 11 可知 $idom(u)$ 也是 $idom(v)$ 的祖先，因此只需证明 $idom(v)$ 支配 $u$。
+
+考虑任意一条 $s$ 到 $u$ 的路径 $P$，我们需要证明 $sdom(u)$ 一定在 $P$ 中。令 $x$ 为 $P$ 中最后一个满足 $x<sdom(u)$ 的节点。如果 $x$ 不存在则必有 $sdom(u)=idom(u) =s$，否则令 $y$ 是 $P$ 中 $x$ 之后在 DFS 树中从 $sdom(u)$ 到 $u$ 的路径上的第一个点。
+
+与定理 2 的证明过程同理，我们可以得到 $sdom(y) \le x$。根据引理 10 有 $sdom(y)\le x<idom(v) \le sdom(v)$。至此，由 $v$ 的定义可知 $y$ 不能是 $sdom(u)$ 的后代；另一方面，$y$ 不能既是 $idom(v)$ 的后代也是 $v$ 的祖先，否则沿 DFS 树从 $s$ 到 $sdom(y)$ 再沿 P 走到 $y$，最后沿 DFS 树走到 $v$ 的这条路径不经过 $idom(v)$，与支配点的定义矛盾。因此 $y=idom(v)$，即 $P$ 包含 $idom(v)$。
+
+根据以上两个定理我们能够得到 $sdom(u)$ 与 $idom(u)$ 之间的关系。
+
+令 $v$ 是满足 $v$ 在 $sdom(u)$ 与 $u$ 之间的结点的所有节点中，$sdom(v)$ 最小的一个节点，那么：
+
+$$
+idom(u) =
+\left\{ 
+\begin{aligned} 
+& sdom(u), &if\ sdom(u) = sdom(v)
+\\
+&idom(v), &otherwise
+\end{aligned}
+\right.
+$$
```

Comment:
```suggestion
$$
idom(u) =
\left\{ 
\begin{aligned} 
& sdom(u), &\text{if}\ sdom(u) = sdom(v)
\\
&idom(v), &\text{otherwise}
\end{aligned}
\right.
$$
```